### ISSUE TITLE
GradientTreeBoost : OnlineRegression 

### ISSUE THREAD

**Issue by olivbrau** [haifengl/smile#691]
**Is your feature request related to a problem? Please describe.**
GradientTreeBoost is a powerfull machine learning algorithm, but it is difficult and painfull to find the good parameters. We have to make multiple attemps, which can be slow.
But there is one parameter that could be analysed differently and efficiently : ntrees (nb of trees)

**Describe the solution you'd like**
It would be nice to adapt the fitting method to allow the caller to test the model, at each iteration, to compare the evolution of RMSE (for ex.) on training dataset and validation dataset to see the effect of ntrees, and then be abble to detect when the model is overfitting.
It would avoid to test with ntrees=100   then ntrees = 200 etc. which is not efficient.
So, in Smile vocabulary, it consists of making GradientTreeBoost an OnlineRegression with update method.

This mechanism could also allow the caller to monitor the progress of the training (UI with progress bar, etc.) and to stop it if too long.

**Comment by haifengl** [comment]
It is more about early stopping than online learning.

**Comment by olivbrau** [comment]
Yes, I was wrong, it is early stopping. Since there is no early stopping possible with Gradient Tree Boost, I thought that OnlineRegression could let the user to his own mechanism.
We can do it with Neural Network (MLP)  : the user makes his own iteration. I think it is very usefull. And also it let the user to stop learning if someting is wrong, since Gradient Tree Boost can take a long time if not carefully parametrized.

**Comment by kklioss** [comment]
We have added an `IterativeAlgorithmController` class in `smile.util` package. You can pass such a controller to `GradientTreeBoost.fit()` method as part of `Options` in addition to an optional test dataset. If test data is provided, we will calculate rich validation metrics and send it to the subscribers. The controller may choose to early stop the training based on the metrics. Please see the test cases for the usage.

**Comment by haifengl** [comment]
This is released in v4.3.0.

**Comment by olivbrau** [comment]
Wonderful. Thanks a lot for your amazing work !
